File: SampleProjects/GoT-Characters-master/Objective-C/GoT/ArticlesRepository.m
//
//  ArticlesRepository.m
//  GoT
//
//  Created by Paciej on 25/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "ArticlesRepository.h"

@interface ArticlesRepository ()
@property(nonatomic, strong, nonnull) NSUserDefaults *defaults;
@property(nonatomic, strong, nonnull) NSMutableSet *articles;
@property(nonatomic, strong, nonnull) NSMutableSet *temporaryArticles;
@end

static NSString *kSavedArticles = @"SavedArticles";

@implementation ArticlesRepository
- (nonnull instancetype)initWithDefaults:(nonnull NSUserDefaults *)defaults {
    self = [super init];
    if (self) {
        self.defaults = defaults;
        self.articles = [self getFavouriteArticlesFromDefaults];
        self.temporaryArticles = [[NSMutableSet alloc] init];
    }
    return self;
}

- (nonnull NSArray *)savedFavouritesArticles {
    return self.articles.allObjects;
}

- (void)saveFavouriteArticle:(nonnull Article *)article {
    [self.articles addObject:article];
    [self saveFavouriteArticlesToDefaults];
}

- (void)removeFavouriteArticle:(nonnull Article *)article {
    [self.articles removeObject:article];
    [self saveFavouriteArticlesToDefaults];
}

- (void)saveFavouriteArticlesToDefaults {
    NSData *archivedArticles =
    [NSKeyedArchiver archivedDataWithRootObject:self.articles];
    [self.defaults setObject:archivedArticles forKey:kSavedArticles];
    [self.defaults synchronize];
}

- (nonnull NSMutableSet *)getFavouriteArticlesFromDefaults {
    NSData *archivedArticles = [self.defaults objectForKey:kSavedArticles];
    NSMutableSet *set =
    [NSKeyedUnarchiver unarchiveObjectWithData:archivedArticles];
    if (!set) {
        set = [[NSMutableSet alloc] init];
    }
    return set;
}

- (BOOL)isFavouriteArticle:(nonnull Article *)article {
    return
    [self.articles
     filteredSetUsingPredicate:[NSPredicate
                                predicateWithFormat:@"identifier == %@",
                                article.identifier]]
    .count > 0; // This could cause a perfromance issue for large data
    // sets!? TODO: verify and find better solution
}

- (void)storeArticlesTemporarily:(nonnull NSArray *)articles {
    [self.temporaryArticles addObjectsFromArray:articles];
}

- (nonnull NSArray *)temporarilyStoredArticles {
    return self.temporaryArticles.allObjects;
}

@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/ArticlesRepository.swift
//
//  ArticlesRepository.swift
//  GoT
//
//  Created by Paciej on 25/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import Foundation

private var kSavedArticles = "SavedArticles"

class ArticlesRepository: NSObject {
    private var defaults: UserDefaults!
    private var articles: Set<AnyHashable> = []
    private var temporaryArticles: Set<AnyHashable> = []

    init(defaults: UserDefaults) {
        super.init()
        
        self.defaults = defaults
        articles = getFavouriteArticlesFromDefaults()
        temporaryArticles = Set<AnyHashable>()
    
    }

    func savedFavouritesArticles() -> [Any] {
        return Array(articles)
    }

    func saveFavouriteArticle(_ article: Article) {
        articles.insert(article)
        saveFavouriteArticlesToDefaults()
    }

    func removeFavouriteArticle(_ article: Article) {
        articles.remove(article)
        saveFavouriteArticlesToDefaults()
    }

    func isFavouriteArticle(_ article: Article) -> Bool {
        return articles.filter { NSPredicate(format: "identifier == %@", article.identifier).evaluate(with: $0) }.count > 0 // This could cause a perfromance issue for large data
        // sets!? TODO: verify and find better solution
    }

    func storeArticlesTemporarily(_ articles: [Any]) {
        temporaryArticles.formUnion(Set(articles.map { $0 as! AnyHashable }))
    }

    func temporarilyStoredArticles() -> [Any] {
        return Array(temporaryArticles)
    }

    func saveFavouriteArticlesToDefaults() {
        let archivedArticles = NSKeyedArchiver.archivedData(withRootObject: articles)
        defaults.set(archivedArticles, forKey: kSavedArticles)
        defaults.synchronize()
    }

    func getFavouriteArticlesFromDefaults() -> Set<AnyHashable> {
        let archivedArticles = defaults.object(forKey: kSavedArticles) as? Data
        var `set`: Set<AnyHashable>? = nil
        if let anArticles = archivedArticles {
            `set` = NSKeyedUnarchiver.unarchiveObject(with: anArticles) as? Set<AnyHashable>
        }
        if `set` == nil {
            `set` = Set<AnyHashable>()
        }
        return `set` ?? []
    }
}